### PR TITLE
Use timeTraceAsyncProfilerBegin for Source span

### LINKED ISSUE #56554: -ftime-trace incorrect header hierarchy when a header contains a template function for his last symbol
Using the repro case below, you will observe in the traces that `2.h` shows up as an inclusion under `1.h` instead of being an inclusion of `main/cpp`. The addition of a dummy forward declaration at the end of `1.h` is a workaround to the problem and will yield the correct result with `2.h` being under `main.cpp`. In both case, `3.h` remains under `2.h` accordingly.

* `main.cpp`
```c++
#include "1.h"
#include "2.h"
int foo();
```
* `1.h`
```hpp
#pragma once
#include <vector>
struct LARGE {};
template <typename T> auto Zero() -> T { return T{}; }
#if defined GUARD
struct guard;
#endif
```
* `2.h`
```hpp
#pragma once
#include "3.h"
#include <map>
struct Bla {};
```
* `3.h`
```hpp
#pragma once
#include <algorithm>
struct Foo {};
```

To produce the incorrect time trace : `clang -ftime-trace main.cpp -c -o main-wrong.o`
To produce the correct time trace : `clang -DGUARD -ftime-trace main.cpp -c -o main-right.o`

### PR DESCRIPTION
**Pull request by atetubou**
This fixes incorrect trace for https://github.com/llvm/llvm-project/issues/56554.

This shows trace like
https://ui.perfetto.dev/#!/?s=aa809778dc50f2b155b062317fa18bbe2bb2fb9175e6282add8121c7c178214e
for the case shown in https://github.com/llvm/llvm-project/issues/83236.

https://github.com/llvm/llvm-project/pull/83778 is preparing PR.

Fix #56554

### CODE REVIEW

**Comment by atetubou** [comment]
@ZequanWu could you take a look this?

File: llvm/lib/Support/TimeProfiler.cpp [hunk 1]
```diff
@@ -102,23 +104,24 @@ struct llvm::TimeTraceProfiler {
...
+                                llvm::function_ref<std::string()> Detail,
+                                bool AsyncEvent = false) {
+    Stack.emplace_back(std::make_unique<TimeTraceProfilerEntry>(
+        ClockType::now(), TimePointType(), std::move(Name), Detail(),
+        AsyncEvent));
+    return Stack.back().get();
   }
 
   void end() {
+    TimeTraceProfilerEntry *E = Stack.back().get();
```
Comment (ZequanWu) [inline llvm/lib/Support/TimeProfiler.cpp:117]: If we extract end(TimeTraceProfilerEntry &E) to a separate function, probably want to move the assertion to here as well

File: llvm/lib/Support/TimeProfiler.cpp [hunk 2]
```diff
@@ -132,15 +135,18 @@ struct llvm::TimeTraceProfiler {
...
-                      [&](const TimeTraceProfilerEntry &Val) {
-                        return Val.Name == E.Name;
+                      [&](const std::unique_ptr<TimeTraceProfilerEntry> &Val) {
+                        return Val->Name == E.Name;
                       })) {
       auto &CountAndTotal = CountAndTotalPerName[E.Name];
       CountAndTotal.first++;
       CountAndTotal.second += Duration;
-    }
+    };
```
Comment (ZequanWu) [inline llvm/lib/Support/TimeProfiler.cpp:144]: Unnecessary change.
```suggestion
    }
```

**Review by ZequanWu — COMMENTED** [review]
IIUC, the approach you choose here is to let `SemaPPCallbacks` control the "entered file stack" and allow it to remove element (which is file) from middle of the internal stack in `TimeTraceProfiler`, but this creates async event which is not designed for this purpose.

Can we let `SemaPPCallbacks` track the last push file into the stack and when exit file, pop all the elements from the stack until we popped the last pushed file?

**Comment by atetubou** [comment]
> IIUC, the approach you choose here is to let `SemaPPCallbacks` control the "entered file stack" and allow it to remove element (which is file) from middle of the internal stack in `TimeTraceProfiler`, but this creates async event which is not designed for this purpose.
> 
> Can we let `SemaPPCallbacks` track the last push file into the stack and when exit file, pop all the elements from the stack until we popped the last pushed file?

As I wrote in https://github.com/llvm/llvm-project/issues/56554#issuecomment-1975812398, file level span and syntax tree level span should be handled asynchronously. So using such implementation produces incorrect trace in other edge cases.

File: llvm/lib/Support/TimeProfiler.cpp (hunk 1 above)
Comment (atetubou) [reply llvm/lib/Support/TimeProfiler.cpp:117]: But this function calls `end(TimeTraceProfilerEntry &E)` and that covers the assertion?

Comment (ZequanWu) [reply llvm/lib/Support/TimeProfiler.cpp:117]: But that's assertion after the call to `Stack.back()`.

**Comment by ZequanWu** [comment]
> > IIUC, the approach you choose here is to let `SemaPPCallbacks` control the "entered file stack" and allow it to remove element (which is file) from middle of the internal stack in `TimeTraceProfiler`, but this creates async event which is not designed for this purpose.
> > Can we let `SemaPPCallbacks` track the last push file into the stack and when exit file, pop all the elements from the stack until we popped the last pushed file?
> 
> As I wrote in [#56554 (comment)](https://github.com/llvm/llvm-project/issues/56554#issuecomment-1975812398), file level span and syntax tree level span should be handled asynchronously. So using such implementation produces incorrect trace in other edge cases.

Can we choose to expose a handle (maybe an unique id for each entry) from `TimeTraceProfiler` instead of exposing the internal entry?

**Comment by atetubou** [comment]
> > > IIUC, the approach you choose here is to let `SemaPPCallbacks` control the "entered file stack" and allow it to remove element (which is file) from middle of the internal stack in `TimeTraceProfiler`, but this creates async event which is not designed for this purpose.
> > > Can we let `SemaPPCallbacks` track the last push file into the stack and when exit file, pop all the elements from the stack until we popped the last pushed file?
> > 
> > 
> > As I wrote in [#56554 (comment)](https://github.com/llvm/llvm-project/issues/56554#issuecomment-1975812398), file level span and syntax tree level span should be handled asynchronously. So using such implementation produces incorrect trace in other edge cases.
> 
> Can we choose to expose a handle (maybe an unique id for each entry) from `TimeTraceProfiler` instead of exposing the internal entry?

I think pointer to `TimeTraceProfilerEntry` is already like handle as definition of the struct is not written in header file. Having another handle for that seems unnecessary indirection layer to me. Or better to change the name of struct?

**Comment by ZequanWu** [comment]
> > > > IIUC, the approach you choose here is to let `SemaPPCallbacks` control the "entered file stack" and allow it to remove element (which is file) from middle of the internal stack in `TimeTraceProfiler`, but this creates async event which is not designed for this purpose.
> > > > Can we let `SemaPPCallbacks` track the last push file into the stack and when exit file, pop all the elements from the stack until we popped the last pushed file?
> > > 
> > > 
> > > As I wrote in [#56554 (comment)](https://github.com/llvm/llvm-project/issues/56554#issuecomment-1975812398), file level span and syntax tree level span should be handled asynchronously. So using such implementation produces incorrect trace in other edge cases.
> > 
> > 
> > Can we choose to expose a handle (maybe an unique id for each entry) from `TimeTraceProfiler` instead of exposing the internal entry?
> 
> I think pointer to `TimeTraceProfilerEntry` is already like handle as definition of the struct is not written in header file. Having another handle for that seems unnecessary indirection layer to me. Or better to change the name of struct?

Okay, make sense. Looks good for me.
Adding @MaskRay to review in case he has something to say.

**Comment by atetubou** [comment]
@MaskRay could you review this?

**Review by ZequanWu — APPROVED** [review]

File: clang/lib/Sema/Sema.cpp [hunk 3]
```diff
@@ -166,8 +167,9 @@ class SemaPPCallbacks : public PPCallbacks {
     }
     case ExitFile:
       if (!IncludeStack.empty()) {
-        if (llvm::timeTraceProfilerEnabled())
-          llvm::timeTraceProfilerEnd();
+        if (llvm::timeTraceProfilerEnabled()) {
+          llvm::timeTraceProfilerEnd(ProfilerStack.pop_back_val());
```
Comment (MaskRay) [inline clang/lib/Sema/Sema.cpp:171]: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
Manually fixed the style and landed as 83eb8aee4bf9d518b3a2b485640207e7717805b4

**Comment by atetubou** [comment]
Thank you for your review and merge!
